Use logging while setting default solver in PDE simulations

santisoler opened the pull request:
#### Summary

Ditch the `DefaultSolverWarning` in `get_default_solver` and make the base PDE simulation class to handle the information message to users about using a default solver. The information message is sent through a new custom logger defined using `logging`. The information is sent from the constructor after obtaining the solver through `get_default_solver`. The solver class gets cached on instantiation, so the simulation will use the same solver even if the default solver changes later on. Warn users regarding poor performance when setting `SolverLU` or `Solver` as the solver. Adjust and extend tests. Also, add a `simpeg.utils.get_logger` function that returns the logger object used across the project, and a new `PerformanceWarning` class in a new `simpeg.utils.warnings` submodule.

#### PR Checklist
* [ ] If this is a work in progress PR, set as a Draft PR
* [ ] Linted my code according to the [style guides](https://docs.simpeg.xyz/latest/content/getting_started/contributing/code-style.html).
* [ ] Added [tests](https://docs.simpeg.xyz/latest/content/getting_started/contributing/testing.html) to verify changes to the code.
* [ ] Added necessary documentation to any new functions/classes following the
      expect [style](https://docs.simpeg.xyz/latest/content/getting_started/contributing/documentation.html).
* [ ] Marked as ready for review (if this is was a draft PR), and converted
      to a Pull Request
* [ ] Tagged ``@simpeg/simpeg-developers`` when ready for review.

#### Reference issue
<!--Example: write "Closes #NNNN" to automatically close that issue on merge.-->

Fixes #1660

#### What does this implement/fix?
<!--Please explain your changes.-->

#### Additional information
<!--Any additional information you think is important.-->


<!--
Once all tests pass and the code has been reviewed and approved, it will be merged into main
-->

jcapriot commented on simpeg/utils/logger.py:
```diff
@@ -0,0 +1,23 @@
+    """
+    Create logger for SimPEG.
+    """
+    logger = logging.getLogger(__name__)
```
This will make it look like every logged line comes from this file. It would make more sense to instantiate a logger for the simulation in my opinion.

santisoler replied:
Make sense. What about a global logger for SimPEG?

jcapriot commented:
I’m still very much in favor of it being a warning. The amount of times we’ve had people run simulations, asking why the runtime is so different, without knowing what solver they used has drastically decreased.

santisoler commented:
> I’m still very much in favor of it being a warning. The amount of times we’ve had people run simulations, asking why the runtime is so different, without knowing what solver they used has drastically decreased.

Right. The idea is not to make that information to disappear, but to avoid warning users, and inform them instead. I think @prisae had a good point by saying that not choosing a solver is the default, so should we warn users if they are using the default choice? An information sounds better suited for that case.

Last week we were also discussing two relevant points:
1. Moving the warning from the `solver` setter to the constructor. So, users get the information when the simulation is instantiated and not when the solver gets accessed.
2. Making it simpler to silent that information/warning. Instructing users to filterwarnings can be quite dangerous. A `verbose` flag (like the one in this PR) seemed a simpler choice.

prisae commented:
I like the approach. And the `StreamHandler` will make sure it is also printed to screen, not just to a potential logging file, so the user does get informed ('warned'). Starting with the logging module is probably good anyway, I do not use it enough myself...

santisoler commented:
I've just applied a few changes to this PR, inline with the conversation we had yesterday. In a nutshell:

* We stream an INFO message through the logger instead of a warning every time the PDE simulation uses the `get_default_solver` to set a solver.
* The solver class gets cached, so the simulation will use the solver set in the constructor, even if later on the default solver changes. This makes the process _less magical_.
* I removed the `verbose` flag. We agreed that it's better to instruct users to pass their own solver if the don't want to see the message again.
* If the `SolverLU` is chosen, users will get a warning about its poor performance, recommending them to change to Pardiso or MUMPS.
* I added logging capabilities by exposing a `get_logger` function in `simpeg.utils`. We can think about using further in the future.
* I also added a `simpeg.utils.warnings` submodule where we can add custom warnings.

santisoler commented on simpeg/base/pde_simulation.py:
```diff
@@ -433,12 +434,24 @@ class BasePDESimulation(BaseSimulation):
+                "To avoid receiving this message, pass a solver to the simulation. "
+                "For example:"
+                "\n\n"
+                "  from simpeg.utils.solver_utils import get_default_solver\n"
```
```suggestion
                "  from simpeg.utils import get_default_solver\n"
```

prisae commented:
Great, thanks @santisoler!
